esp-hal/src/spi/master.rs

```diff
@@ -1449,270 +1387,486 @@ pub mod dma {
+                    State::InUse => unreachable!(),
+                }
+                match take(&mut self.state) {
+                    State::Idle(spi, tx_buf, rx_buf) => (spi, tx_buf, rx_buf),
+                    State::Reading(transfer, tx_buf) => {
+                        let (spi, rx_buf) = transfer.wait();
+                        (spi, tx_buf, rx_buf)
                     }
-                    future.await?;
-
-                    self.spi.flush()?;
+                    State::Writing(transfer, rx_buf) => {
+                        let (spi, tx_buf) = transfer.wait();
+                        (spi, tx_buf, rx_buf)
+                    }
+                    State::Transferring(transfer) => {
+                        let (spi, (tx_buf, rx_buf)) = transfer.wait();
+                        (spi, tx_buf, rx_buf)
+                    }
+                    State::InUse => unreachable!(),
                 }
-
-                Ok(())
             }
+        }
 
-            async fn transfer(&mut self, read: &mut [u8], write: &[u8]) -> Result<(), Self::Error> {
-                let mut idx = 0;
-                loop {
-                    let write_idx = isize::min(idx, write.len() as isize);
-                    let write_len = usize::min(write.len() - idx as usize, MAX_DMA_SIZE);
-
-                    let read_idx = isize::min(idx, read.len() as isize);
-                    let read_len = usize::min(read.len() - idx as usize, MAX_DMA_SIZE);
-
-                    let mut tx_future = crate::dma::asynch::DmaTxFuture::new(&mut self.channel.tx);
-                    let mut rx_future = crate::dma::asynch::DmaRxFuture::new(&mut self.channel.rx);
-
-                    unsafe {
-                        self.spi.start_transfer_dma(
-                            &mut self.tx_chain,
-                            &mut self.rx_chain,
-                            write.as_ptr().offset(write_idx),
-                            write_len,
-                            read.as_mut_ptr().offset(read_idx),
-                            read_len,
-                            tx_future.tx(),
-                            rx_future.rx(),
-                        )?;
-                    }
-                    let (tx_res, rx_res) = embassy_futures::join::join(tx_future, rx_future).await;
-                    tx_res?;
-                    rx_res?;
-
-                    self.spi.flush()?;
+        impl<'d, T, C> ErrorType for SpiDmaAsyncBus<'d, T, C>
+        where
+            T: InstanceDma,
+            C: DmaChannel,
+            C::P: SpiPeripheral,
+        {
+            type Error = Error;
+        }
 
-                    idx += MAX_DMA_SIZE as isize;
-                    if idx >= write.len() as isize && idx >= read.len() as isize {
-                        break;
-                    }
+        impl<'d, T, C> embedded_hal_async::spi::SpiBus for SpiDmaAsyncBus<'d, T, C>
```

Comment:
I know it wasn't like that before but in general we want to provide an inherent API and have ecosystem traits optionally implemented - since you are touching all of this anyways it's probably a chance to do it now?

Reply:
But not really required in this PR - I understood you planned follow up PRs already

Reply:
Sure, I can do this in the PR that removes asymmetric transfers from the core API. (This makes it a bit easier for me)

We also need to decide what the inherent API should look like, copying the e-hal trait seems lazy to me but I can't think of anything else

Reply:
> ... copying the e-hal trait seems lazy to me but I can't think of anything else

I guess that's good enough 👍